content/inside-rust/stage0-redesign/index.md

```diff
@@ -0,0 +1,70 @@
+| `./x {check,build,test} library`           | Checks/builds/tests in-tree library                         | Check/build/tests stage 1 library                        |
+| `./x {check,build,test} library --stage 0` | Checks/builds/tests in-tree library                         | No-op in case of build, checks/tests precompiled library |
+| `./x {check,build,test} library --stage 1` | Builds in-tree library, checks/builds/tests stage 1 library | Check/build/tests stage 1 library                        |
+
+For `profile = "compiler"`:
+
+| Invocation                                 | Before stage 0 std redesign                                 | After stage 0 std redesign                               |
+|--------------------------------------------|-------------------------------------------------------------|----------------------------------------------------------|
+| `./x check library`                        | Checks in-tree library                                      | Check/build/tests stage 1 library                        |
+| `./x {build,test} library`                 | Builds in-tree library, builds/tests stage 1 library        | Check/build/tests stage 1 library                        |
+| `./x {check,build,test} library --stage 0` | Checks/builds/tests in-tree library                         | No-op in case of build, checks/tests precompiled library |
+| `./x {check,build,test} library --stage 1` | Builds in-tree library, checks/builds/tests stage 1 library | Check/build/tests stage 1 library                        |
+
+For `profile = "tools"`, by default not affected if `download-rustc` is enabled.
+
+## What does this mean for a typical library workflow?
+
+- Crucially, `./x {build,test,check} library --stage 0` becomes no-op and are no longer supported. Building the in-tree std now requires a stage 1 compiler.
+    - Consequently, library contributors are *strongly* encouraged to enable `rust.download-rustc = "if-unchanged"` to avoid having to build a stage 1 compiler. Note that this is the default for `profile = "library"`, but you may need to specify it manually if you don't use a `profile`.
+- `cfg(bootstrap)` should no longer be needed for library sources.
+
+## What does this mean for a typical compiler workflow?
+
+- If unstable library features are being dogfooded, some additional `cfg(bootstrap)` usages may now be needed in compiler sources.
+
+## Why are we making this change?
+
+The previous stage 0 bootstrapping sequence was a source of endless confusion for compiler, library and bootstrap contributors alike, because std had to support being built by *both* a previous beta rustc and in-tree rustc, with `cfg(bootstrap)` in std sources necessary to distinguish between them. By adjusting the stage 0 bootstrap sequence to instead use a precompiled stage 0 std instead of building the in-tree std, we hope to:
```

Comment:
I have an alternative description of the motivation that I think more accurately captures the effect on combined compiler/library changes. Feel free to merge and integrate or so whatever with it that you want:

The compiler and standard library are tightly coupled. Intrinsics and lang items form a broad interface between the two, and whenever it is changed, both sides need to adjust. But the standard library supports being compiled with two different compilers, the in-tree compiler and the stage0 (beta) compiler. This means that all such changes require `cfg(bootstrap)` to support both compilers, which results in a lot of churn of anyone wanting to improve things that touch this area.

Reply:
Yeah, I think that's a way better motivation summary. I'll need to think about how to reword this section.

Reply:
I tried to word the motivation as something to that effect, still not very convincing IMO. (round 4).